extensions/amp-font/0.1/amp-font.js

```diff
@@ -0,0 +1,327 @@
+
+/** @private @const {string} */
+const DEFAULT_VARIANT_ = 'normal';
+
+/** @private @const {string} */
+const DEFAULT_STYLE_ = 'normal';
+
+/** @private @const {string} */
+const DEFAULT_SIZE_ = 'medium';
+
+
+export class AmpFont extends AMP.BaseElement {
+
+
+  /** @override */
+  prerenderAllowed() {
+    return true;
+  }
+
+
+  /** @override */
+  buildCallback() {
+    this.customFontFamily_ = this.element.getAttribute('font-family');
+    this.customFontWeight_ =
+        this.element.getAttribute('font-weight') || DEFAULT_WEIGHT_;
+    this.customFontStyle_ =
+        this.element.getAttribute('font-style') || DEFAULT_STYLE_;
+    this.customFontVariant_ =
+        this.element.getAttribute('font-variant') || DEFAULT_VARIANT_;
+    this.timeoutHandle_ = null;
+    this.hasTimedOut_ = false;
+    this.domElements_ = {
+      CONTAINER: null,
+      DEFAULT_FONT_ELEMENTS: [],
+      CUSTOM_FONT_ELEMENT: null
+    };
+    this.win_ = this.getWin();
+    this.document_ = this.win_.document;
+    this.documentElement_ = this.document_.documentElement;
+    this.startTime_ = timer.now();
+    this.startLoad_();
+  }
+
+
+  /**
+   * Start to download the font if it is already not available.
+   * @private
+   */
+  startLoad_() {
+    this.beginTimeout_();
+    if (this.canUseNativeApis_()) {
+      this.loadFontWithNativeApis_();
+    } else {
+      this.forceLoadFont_();
+    }
+  }
+
+  /**
+   * @private
+   */
+  canUseNativeApis_() {
+    return 'fonts' in this.document_;
+  }
+
+
+  /**
+   * If the browser supports font loading events use them to load the font.
+   * @private
+   */
+  loadFontWithNativeApis_() {
+    /* style | variant | weight | size/line-height | family */
+    /* font: italic small-caps bolder 16px/3 cursive; */
+    let fontString = (
+      this.customFontStyle_ + ' ' +
+      this.customFontVariant_ + ' ' +
+      this.customFontWeight_ + ' ' +
+      DEFAULT_SIZE_ + ' ' +
+      this.customFontFamily_);
+    if (this.document_.fonts.check(fontString)) {
+      this.onFontLoadSuccess_();
+    } else {
+      this.document_.fonts.load(fontString).then(() => {
+        this.onFontLoadSuccess_();
+      });
+    }
+  }
+
+
+  /**
+   * Make the browsers that don't support font loading events to download the
+   * custom font by creating an element (with text) not visible on the viewport.
+   * font download is detected by comparing the elements height and width with
+   * measurements between default fonts and custom font.
+   * @private
+   */
+  forceLoadFont_() {
+    // Create DOM elements
+    this.createElements_();
+    // Measure until timeout (or font load).
+    let vsyncTask = {
+      measure: () => {
+        if (this.compareMeasurements_()) {
+          this.onFontLoadSuccess_();
+        } else if (this.hasTimedOut_) {
+          this.onFontLoadError_();
+        } else {
+          vsync.run(vsyncTask);
+        }
+      },
+      mutate: () => {}
+    };
+    vsync.run(vsyncTask);
```

Comment:
Please do the following:

1. Create a bug that explains why vsync might be a problem here when document is not visible.
2. Add TODO here that would reference this new bug# with a short summary.